---
title: "Report Sleep Deprivation"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
    html_document:
        fig_height: 9
        fig_width: 9
        toc: yes
        toc_float: yes
        toc_depth: 4
        code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r include=FALSE}
knitr::opts_chunk$set(eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE,
                      fig.width=9, fig.height=9, cache=TRUE)
source("../../R/reportIncludes.R")
source("../../R/luciaVolcanoFunctions.R")
library(plotly)
library(RUVSeq)
```

# Importing data

```{r eval=TRUE, echo=TRUE, warning=FALSE, message=TRUE}
countMatrix <- ReadDataFrameFromTsv(file.name.path="../../data/refSEQ_countMatrix.txt")
# head(countMatrix)

designMatrix <- ReadDataFrameFromTsv(file.name.path="../../design/all_samples_short_names_noRS2HC7.tsv")
# head(designMatrix)

# rownames <- as.character(rownames(countMatrix))

# rownames <- rownames[order(rownames)]
# rownames.map <- convertGenesViaBiomart(specie="mm10", filter="entrezgene",
#                     filter.values=rownames, attrs=c("external_gene_name",
#                                 "mgi_symbol", "entrezgene"))
# 
# noNaCountMatrix <- attachGeneColumnToDf(mainDf=countMatrix,
#                                 genesMap=rownames.map,
#                                 rowNamesIdentifier="entrezgene",
#                                 mapFromIdentifier="entrezgene",
#                                 mapToIdentifier="external_gene_name")


filteredCountsProp <- filterLowCounts(counts.dataframe=countMatrix, 
                                    is.normalized=FALSE,
                                    design.dataframe=designMatrix,
                                    cond.col.name="gcondition",
                                    method.type="Proportion")
```


# Plot PCA of log unnormalized data

```{r eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}
PlotPCAPlotlyFunction(counts.data.frame=log1p(filteredCountsProp), design.matrix=designMatrix, shapeColname="condition", colorColname="genotype", xPCA="PC1", yPCA="PC2", plotly.flag=TRUE, show.plot.flag=TRUE, prefix.plot="Prop-Un-Norm")
# pc1_2 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(filteredCountsProp), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC1", yPCA="PC2", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="Prop-Un-Norm")
# pc2_3 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(filteredCountsProp), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC2", yPCA="PC3", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="Prop-Un-Norm")
# pc1_3 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(filteredCountsProp), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC1", yPCA="PC3", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="Prop-Un-Norm")
# plotly::subplot(pc1_2, pc2_3, pc1_3, nrows=2, margin = 0.1, titleX=TRUE, titleY=TRUE)
```


# control genes

Loading Negative Control Genes to normalize data
```{r eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}

library(readxl)

sd.ctrls <- read_excel(path="../../data/controls/Additional File 4 full list of BMC genomics SD&RS2.xlsx", sheet=1)
sd.ctrls <- sd.ctrls[order(sd.ctrls$adj.P.Val),]

sd.neg.ctrls <- sd.ctrls[sd.ctrls$adj.P.Val > 0.9, ]

sd.neg.ctrls <- sd.neg.ctrls$`MGI Symbol`
sd.neg.ctrls <- sd.neg.ctrls[-which(is.na(sd.neg.ctrls))]

int.neg.ctrls <- sd.neg.ctrls
int.neg.ctrls <- unique(int.neg.ctrls)
neg.map <- convertGenesViaMouseDb(gene.list=int.neg.ctrls, fromType="SYMBOL",
                                "ENTREZID")
sum(is.na(neg.map$ENTREZID))
neg.ctrls.entrez <- as.character(neg.map$ENTREZID)

# neg.map <- convertGenesViaBiomart(specie="mm10", filter="mgi_symbol",
#                     filter.values=int.neg.ctrls, c("external_gene_name",
#                                 "mgi_symbol", "entrezgene"))
# 
# neg.map.nna <- neg.map[-which(is.na(neg.map$entrezgene)),]
# neg.ctrls.entrez <- as.character(neg.map.nna$entrezgene)

ind.ctrls <- which(rownames(filteredCountsProp) %in% neg.ctrls.entrez)
counts.neg.ctrls <- filteredCountsProp[ind.ctrls,]
# pc1_2 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(counts.neg.ctrls), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC1", yPCA="PC2", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="Neg Ctrls not Norm")
# pc2_3 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(counts.neg.ctrls), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC2", yPCA="PC3", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="Neg Ctrls not Norm")
# pc1_3 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(counts.neg.ctrls), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC1", yPCA="PC3", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="Neg Ctrls not Norm")
# plotly::subplot(pc1_2, pc2_3, pc1_3, nrows=2, margin = 0.1, titleX=TRUE, titleY=TRUE)
```

## positive control genes

Positive Control Genes

```{r}
## sleep deprivation
sd.lit.pos.ctrls <- read_excel("../../data/controls/SD_RS_PosControls_final.xlsx", 
                        sheet=1)
colnames(sd.lit.pos.ctrls) <- sd.lit.pos.ctrls[1,]
sd.lit.pos.ctrls <- sd.lit.pos.ctrls[-1,]


sd.est.pos.ctrls <- read_excel("../../data/controls/SD_RS_PosControls_final.xlsx", 
                        sheet=3)

sd.pos.ctrls <- cbind(sd.est.pos.ctrls$`MGI Symbol`, "est")
sd.pos.ctrls <- rbind(sd.pos.ctrls, cbind(sd.lit.pos.ctrls$Gene, "lit"))

sd.pos.ctrls <- sd.pos.ctrls[-which(duplicated(sd.pos.ctrls[,1])),]
sd.pos.ctrls <- sd.pos.ctrls[-which(is.na(sd.pos.ctrls[,1])),]

```


# Normalizations

## TMM Normalization

```{r eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}

normPropCountsUqua <- NormalizeData(data.to.normalize=filteredCountsProp, 
                                    norm.type="tmm", 
                                    design.matrix=designMatrix)

PlotPCAPlotlyFunction(counts.data.frame=log1p(normPropCountsUqua), 
                    design.matrix=designMatrix, shapeColname="condition", 
                    colorColname="genotype", xPCA="PC1", yPCA="PC2", 
                    plotly.flag=TRUE, show.plot.flag=TRUE, 
                    prefix.plot="TMM-Norm")

# pc1_2 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(normPropCountsUqua), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC1", yPCA="PC2", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="TMM-Norm")
# pc2_3 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(normPropCountsUqua), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC2", yPCA="PC3", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="TMM-Norm")
# pc1_3 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(normPropCountsUqua), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC1", yPCA="PC3", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="TMM-Norm")
# plotly::subplot(pc1_2, pc2_3, pc1_3, nrows=2, margin = 0.1, titleX=TRUE, titleY=TRUE)


pal <- RColorBrewer::brewer.pal(9, "Set1")
plotRLE(as.matrix(normPropCountsUqua), outline=FALSE, col=pal[designMatrix$gcondition])

```

## TMM + RUVs Normalization

K=5

```{r eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}
library(RUVSeq)
neg.ctrl.list <- rownames(counts.neg.ctrls)
groups <- makeGroups(designMatrix$gcondition)
ruvedSExprData <- RUVs(as.matrix(round(normPropCountsUqua)), cIdx=neg.ctrl.list,
                       scIdx=groups, k=5)

normExprData <- ruvedSExprData$normalizedCounts

PlotPCAPlotlyFunction(counts.data.frame=log1p(normExprData), 
                    design.matrix=designMatrix, shapeColname="condition", 
                    colorColname="genotype", xPCA="PC1", yPCA="PC2", 
                    plotly.flag=FALSE, show.plot.flag=TRUE, save.plot=TRUE, 
                    plot.folder="aaa",
                    prefix.plot=NULL)

# pc1_2 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(normExprData), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC1", yPCA="PC2", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="TMM+RUV-Norm")
# pc2_3 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(normExprData), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC2", yPCA="PC3", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="TMM+RUV-Norm")
# pc1_3 <- PlotPCAPlotlyFunction(counts.data.frame=log1p(normExprData), design.matrix=designMatrix, shapeColname="genotype", colorColname="condition", xPCA="PC1", yPCA="PC3", plotly.flag=TRUE, show.plot.flag=FALSE, prefix.plot="TMM+RUV-Norm")
# plotly::subplot(pc1_2, pc2_3, pc1_3, nrows=2, margin = 0.1, titleX=TRUE, titleY=TRUE)

pal <- RColorBrewer::brewer.pal(9, "Set1")
plotRLE(normExprData, outline=FALSE, col=pal[designMatrix$gcondition])
```

# edgeR DE Analysis

```{r}
### edgering
padj.thr <- 0.05
desMat <- cbind(designMatrix, ruvedSExprData$W)
colnames(desMat) <- c(colnames(designMatrix), colnames(ruvedSExprData$W))

cc <- c("WTSD5 - WTHC5", "KOHC5 - WTHC5",
        "KOSD5 - WTSD5", "KOSD5 - KOHC5")

rescList1 <- applyEdgeR(counts=filteredCountsProp, design.matrix=desMat,
                        factors.column="gcondition", 
                        weight.columns=c("W_1", "W_2", "W_3", "W_4", "W_5"),
                        contrasts=cc, useIntercept=FALSE, p.threshold=1,
                        is.normalized=FALSE, verbose=TRUE)
names <- names(rescList1)
rescList1 <- lapply(seq_along(rescList1), function(i) 
{
    attachMeans(normalized.counts=normExprData, design.matrix=desMat, 
                factor.column="gcondition", contrast.name=names(rescList1)[i],
                de.results=rescList1[[i]])
})
names(rescList1) <- names

```

## WTSD5 - WTHC5

### pvalue histogram

```{r}

PlotHistPvalPlot(de.results=rescList1[[1]], design.matrix=desMat, 
                show.plot.flag=TRUE, plotly.flag=TRUE, 
                prefix.plot=names(rescList1)[1])
```

### volcano plot

```{r eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}
## mapping ensembl gene id using biomart
# sd.pos.ctrls <- sd.pos.ctrls$`MGI Symbol`
# sd.pos.ctrls <- sd.pos.ctrls[-which(is.na(sd.pos.ctrls))]
# length(sd.pos.ctrls)

# res.o.map <- convertGenesViaBiomart(specie="mm10", filter="entrezgene",
#                             filter.values=rownames(rescList1[[1]]),
#                             c("external_gene_name", "mgi_symbol", "entrezgene"))
# 
# res.o <- attachGeneColumnToDf(mainDf=rescList1[[1]],
#                                 genesMap=res.o.map,
#                                 rowNamesIdentifier="entrezgene",
#                                 mapFromIdentifier="entrezgene",
#                                 mapToIdentifier="external_gene_name")


res.o.map1 <- convertGenesViaMouseDb(gene.list=rownames(rescList1[[1]]), 
                            fromType="ENTREZID")

res.o <- attachGeneColumnToDf(mainDf=rescList1[[1]],
                                genesMap=res.o.map1,
                                rowNamesIdentifier="ENTREZID",
                                mapFromIdentifier="ENTREZID",
                                mapToIdentifier="SYMBOL")


WriteDataFrameAsTsv(data.frame.to.save=res.o, 
                    file.name.path=paste0(names(rescList1)[1], "_edgeR"))

vp <- luciaVolcanoPlot(res.o, sd.pos.ctrls, prefix=names(rescList1)[1], 
                    threshold=padj.thr)
ggplotly(vp)

# vp <- PlotVolcanoPlot(de.results=res.o, counts.dataframe=normExprData, 
#                 design.matrix=desMat,
#                 show.plot.flag=FALSE, plotly.flag=FALSE, save.plot=FALSE, 
#                 prefix.plot=names(rescList1)[1], threshold=padj.thr)#, 
#                 #positive.ctrls.list=sd.pos.ctrls)

de <- sum(res.o$FDR < padj.thr)
nde <- sum(res.o$FDR >= padj.thr)
detable <- cbind(de,nde)
rownames(detable) <- names(rescList1)[1]
ddetable <- detable

tot.ctrls <- dim(sd.pos.ctrls)[1]
idx.pc <- which(tolower(res.o$gene) %in% tolower(sd.pos.ctrls[,1]))
tot.pc.de <- sum(res.o$FDR[idx.pc] < padj.thr)
tot.pc.nde <- length(idx.pc) - tot.pc.de
pos.df <- cbind(tot.ctrls, tot.pc.de, tot.pc.nde)
colnames(pos.df) <- c("total_p.ctrl", "p.ctrl_de_mapped", 
                    "p.ctrl_notde_mapped")
rownames(pos.df) <- names(rescList1)[1]

wt <- res.o[which(res.o$FDR < 0.05),]
wt.sign.genes.entrez <- rownames(res.o)[which(res.o$FDR < 0.05)]

# PlotMAPlotCounts(de.results=res.o, counts.dataframe=normExprData, design.matrix=desMat,
#                  show.plot.flag=TRUE, plotly.flag=TRUE, save.plot=FALSE, prefix.plot=names(rescList1)[1], threshold=0.05)
```

## KOHC5 - WTHC5

### pvalue histogram

```{r}
PlotHistPvalPlot(de.results=rescList1[[2]], design.matrix=desMat, 
                show.plot.flag=TRUE, plotly.flag=TRUE, 
                prefix.plot=names(rescList1)[2])
```

### volcano plot

```{r eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}
# rs2.o.map <- convertGenesViaBiomart(specie="mm10", filter="entrezgene",
#                             filter.values=rownames(rescList1[[2]]),
#                             c("external_gene_name", "mgi_symbol", "entrezgene"))
# 
# res.rs2.o <- attachGeneColumnToDf(mainDf=rescList1[[2]],
#                                 genesMap=rs2.o.map,
#                                 rowNamesIdentifier="entrezgene",
#                                 mapFromIdentifier="entrezgene",
#                                 mapToIdentifier="external_gene_name")
rs2.o.map <- convertGenesViaMouseDb(gene.list=rownames(rescList1[[2]]), 
                            fromType="ENTREZID")

res.rs2.o <- attachGeneColumnToDf(mainDf=rescList1[[2]],
                                genesMap=rs2.o.map,
                                rowNamesIdentifier="ENTREZID",
                                mapFromIdentifier="ENTREZID",
                                mapToIdentifier="SYMBOL")
WriteDataFrameAsTsv(data.frame.to.save=res.rs2.o, 
                    file.name.path=paste0(names(rescList1)[2], "_edgeR"))

vp <- luciaVolcanoPlot(res.rs2.o, positive.controls.df=NULL, 
                       prefix=names(rescList1)[2], 
                       threshold=padj.thr)
ggplotly(vp)

# PlotVolcanoPlot(de.results=res.rs2.o, counts.dataframe=normExprData, 
#                 design.matrix=desMat,
#                 show.plot.flag=TRUE, plotly.flag=TRUE, save.plot=FALSE, 
#                 prefix.plot=names(rescList1)[2], threshold=padj.thr, 
#                 positive.ctrls.list=rs2.pos.ctrls)
de <- sum(res.rs2.o$FDR < padj.thr)
nde <- sum(res.rs2.o$FDR >= padj.thr)
detable <- cbind(de,nde)
rownames(detable) <- names(rescList1)[2]
ddetable <- rbind(ddetable, detable)

kowthc5 <- res.rs2.o[which(res.rs2.o$FDR < 0.05),]
kowthc5.sign.genes.entrez <- rownames(res.rs2.o)[which(res.rs2.o$FDR < 0.05)]

# 
# tot.ctrls <- dim(rs.pos.ctrls)[1]
# idx.pc <- which(tolower(res.o$gene) %in% tolower(rs.pos.ctrls[,1]))
# tot.pc.de <- sum(res.o$FDR[idx.pc] < padj.thr)
# tot.pc.nde <- length(idx.pc) - tot.pc.de
# pos.dff <- cbind(tot.ctrls, tot.pc.de, tot.pc.nde)
# rownames(pos.dff) <- names(rescList1)[2]
# pos.df <- rbind(pos.df, pos.dff)
# # colnames(pos.df) <- c("total p.ctrl", "p.ctrl de mapped", 
# #                     "p.ctrl not de mapped")

# PlotMAPlotCounts(de.results=res.o, counts.dataframe=normExprData, design.matrix=desMat,
                 # show.plot.flag=TRUE, plotly.flag=TRUE, save.plot=FALSE, prefix.plot=names(rescList1)[2], threshold=0.05)
```


## KOSD5 - WTSD5

### pvalue histogram

```{r}

PlotHistPvalPlot(de.results=rescList1[[3]], design.matrix=desMat, 
                show.plot.flag=TRUE, plotly.flag=TRUE, 
                prefix.plot=names(rescList1)[3])
```

### volcano plot

```{r eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}
## mapping ensembl gene id using biomart

# res.o.map <- convertGenesViaBiomart(specie="mm10", filter="entrezgene",
#                             filter.values=rownames(rescList1[[3]]),
#                             c("external_gene_name", "mgi_symbol", "entrezgene"))
# 
# res.o <- attachGeneColumnToDf(mainDf=rescList1[[3]],
#                                 genesMap=res.o.map,
#                                 rowNamesIdentifier="entrezgene",
#                                 mapFromIdentifier="entrezgene",
#                                 mapToIdentifier="external_gene_name")


res.o.map <- convertGenesViaMouseDb(gene.list=rownames(rescList1[[2]]), 
                            fromType="ENTREZID")

res.o <- attachGeneColumnToDf(mainDf=rescList1[[3]],
                                genesMap=res.o.map,
                                rowNamesIdentifier="ENTREZID",
                                mapFromIdentifier="ENTREZID",
                                mapToIdentifier="SYMBOL")

WriteDataFrameAsTsv(data.frame.to.save=res.o, 
                    file.name.path=paste0(names(rescList1)[3], "_edgeR"))

vp <- luciaVolcanoPlot(res.o, positive.controls.df=sd.pos.ctrls, 
                        prefix=names(rescList1)[3], threshold=padj.thr)
ggplotly(vp)


de <- sum(res.o$FDR < padj.thr)
nde <- sum(res.o$FDR >= padj.thr)
detable <- cbind(de,nde)
rownames(detable) <- names(rescList1)[3]
ddetable <- rbind(ddetable, detable)


tot.ctrls <- dim(sd.pos.ctrls)[1]
idx.pc <- which(tolower(res.o$gene) %in% tolower(sd.pos.ctrls[,1]))
tot.pc.de <- sum(res.o$FDR[idx.pc] < padj.thr)
tot.pc.nde <- length(idx.pc) - tot.pc.de
pos.dff <- cbind(tot.ctrls, tot.pc.de, tot.pc.nde)
rownames(pos.dff) <- names(rescList1)[3]
pos.df <- rbind(pos.df, pos.dff)

kowtsd5 <- res.o[which(res.o$FDR < 0.05),]
kowtsd5.sign.genes.entrez <- rownames(res.o)[which(res.o$FDR < 0.05)]

# PlotMAPlotCounts(de.results=res.o, counts.dataframe=normExprData, design.matrix=desMat,
#                  show.plot.flag=TRUE, plotly.flag=TRUE, save.plot=FALSE, prefix.plot=names(rescList1)[1], threshold=0.05)

```

## KOSD5 - WTHC5

### pvalue histogram

```{r}

PlotHistPvalPlot(de.results=rescList1[[4]], design.matrix=desMat, 
                show.plot.flag=TRUE, plotly.flag=TRUE, 
                prefix.plot=names(rescList1)[4])
```

### volcano plot

```{r eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}
## mapping ensembl gene id using biomart

# res.o.map <- convertGenesViaBiomart(specie="mm10", filter="entrezgene",
#                             filter.values=rownames(rescList1[[4]]),
#                             c("external_gene_name", "mgi_symbol", "entrezgene"))
# 
# 
# WriteDataFrameAsTsv(data.frame.to.save=rescList1[[4]], 
#                     file.name.path=paste0(names(rescList1)[4], "_edgeR"))
# 
# res.o <- attachGeneColumnToDf(mainDf=rescList1[[4]],
#                                 genesMap=res.o.map,
#                                 rowNamesIdentifier="entrezgene",
#                                 mapFromIdentifier="entrezgene",
#                                 mapToIdentifier="external_gene_name")



res.o.map <- convertGenesViaMouseDb(gene.list=rownames(rescList1[[4]]), 
                            fromType="ENTREZID")

res.o <- attachGeneColumnToDf(mainDf=rescList1[[4]],
                                genesMap=res.o.map,
                                rowNamesIdentifier="ENTREZID",
                                mapFromIdentifier="ENTREZID",
                                mapToIdentifier="SYMBOL")

WriteDataFrameAsTsv(data.frame.to.save=res.o, 
                    file.name.path=paste0(names(rescList1)[4], "_edgeR"))


vp <- luciaVolcanoPlot(res.o, positive.controls.df=sd.pos.ctrls, 
                        prefix=names(rescList1)[4], threshold=padj.thr)
ggplotly(vp)
# 
# PlotVolcanoPlot(de.results=res.o, counts.dataframe=normExprData, 
#                 design.matrix=desMat,
#                 show.plot.flag=TRUE, plotly.flag=TRUE, save.plot=FALSE, 
#                 prefix.plot=names(rescList1)[4], threshold=padj.thr, 
#                 positive.ctrls.list=NULL)

de <- sum(res.o$FDR < padj.thr)
nde <- sum(res.o$FDR >= padj.thr)
detable <- cbind(de,nde)
rownames(detable) <- names(rescList1)[4]
ddetable <- rbind(ddetable, detable)


tot.ctrls <- dim(sd.pos.ctrls)[1]
idx.pc <- which(tolower(res.o$gene) %in% tolower(sd.pos.ctrls[,1]))
tot.pc.de <- sum(res.o$FDR[idx.pc] < padj.thr)
tot.pc.nde <- length(idx.pc) - tot.pc.de
pos.dff <- cbind(tot.ctrls, tot.pc.de, tot.pc.nde)
rownames(pos.dff) <- names(rescList1)[2]
pos.df <- rbind(pos.df, pos.dff)

ko <- res.o[which(res.o$FDR < 0.05),]
ko.sign.genes.entrez <- rownames(res.o)[which(res.o$FDR < 0.05)]

# PlotMAPlotCounts(de.results=res.o, counts.dataframe=normExprData, design.matrix=desMat,
#                  show.plot.flag=TRUE, plotly.flag=TRUE, save.plot=FALSE, prefix.plot=names(rescList1)[1], threshold=0.05)
```

## DE TABLE + Positive Controls table

```{r}
ddetable
pos.df
```


# Venn KOHC5-WTHC5 vs KOSD5-WTSD5

```{r}
source("../../R/venn2.R")
gene.map <- convertGenesViaMouseDb(gene.list=rownames(normExprData),
                                   fromType="ENTREZID", toType="SYMBOL")
venn <- Venn2de(x=kowthc5.sign.genes.entrez, y=kowtsd5.sign.genes.entrez, 
        label1="KOHC5_WTHC5", label2="KOSD5_WTSD5",
        title="KOHC5_WTHC5 venn KOSD5_WTSD5", plot.dir="./",
        conversion.map=gene.map)



```


## plot genes in KOWT-SD5 but NOT in KOWT-HC5
```{r}
source("../../R/plotGeneProfile.R")
ynox <- venn$YnoX
ynoxdf <- subsetDfByCol(kowtsd5, ynox$ba)

ynoxdf <- attachGeneColumnToDf(mainDf=ynoxdf, genesMap=gene.map, 
                                  rowNamesIdentifier="ENTREZID", 
                                  mapFromIdentifier="ENTREZID", 
                                  mapToIdentifier="SYMBOL")
ynoxdf <- ynoxdf[order(ynoxdf$FDR),]

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=ynoxdf$gene[1],
            res.o=ynoxdf, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=ynoxdf$gene[2],
            res.o=ynoxdf, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=ynoxdf$gene[3],
            res.o=ynoxdf, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=ynoxdf$gene[4],
            res.o=ynoxdf, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=ynoxdf$gene[5],
            res.o=ynoxdf, show.plot=TRUE, plotly.flag=TRUE)

```

## plot genes in KOWT-HC5 but NOT in KOWT-SD5
```{r}
source("../../R/plotGeneProfile.R")
xnoy <- venn$XnoY
xnoydf <- subsetDfByCol(kowthc5, xnoy$ab)

xnoydf <- attachGeneColumnToDf(mainDf=xnoydf, genesMap=gene.map, 
                                  rowNamesIdentifier="ENTREZID", 
                                  mapFromIdentifier="ENTREZID", 
                                  mapToIdentifier="SYMBOL")
xnoydf <- xnoydf[order(xnoydf$FDR),]

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=xnoydf$gene[1],
            res.o=xnoydf, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=xnoydf$gene[2],
            res.o=xnoydf, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=xnoydf$gene[3],
            res.o=xnoydf, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=xnoydf$gene[4],
            res.o=xnoydf, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=xnoydf$gene[5],
            res.o=xnoydf, show.plot=TRUE, plotly.flag=TRUE)

```


## plot genes in intersection
```{r}
source("../../R/plotGeneProfile.R")
int.genes <- intersect(kowthc5.sign.genes.entrez, kowtsd5.sign.genes.entrez)
int.genes <- as.data.frame(int.genes)
rownames(int.genes) <- int.genes[,1]
int.genes <- attachGeneColumnToDf(mainDf=int.genes, genesMap=gene.map, 
                                  rowNamesIdentifier="ENTREZID", 
                                  mapFromIdentifier="ENTREZID", 
                                  mapToIdentifier="SYMBOL")

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=int.genes$gene[1],
            res.o=int.genes, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=int.genes$gene[2],
            res.o=int.genes, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=int.genes$gene[3],
            res.o=int.genes, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=int.genes$gene[4],
            res.o=int.genes, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData, 
            design.matrix=designMatrix, 
            gene.name=int.genes$gene[5],
            res.o=int.genes, show.plot=TRUE, plotly.flag=TRUE)
```

## Special genes profiles

```{r}
colnames(gene.map) <- c("GeneID", "gene")
rownames(gene.map) <- gene.map$GeneID
source("../../R/plotGeneProfile.R")

geneProfileLucia(normalized.counts=normExprData,
            design.matrix=designMatrix,
            gene.name="Hdac7",
            res.o=gene.map, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData,
            design.matrix=designMatrix,
            gene.name="Jun",
            res.o=gene.map, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData,
            design.matrix=designMatrix,
            gene.name="Per3",
            res.o=gene.map, show.plot=TRUE, plotly.flag=TRUE)

geneProfileLucia(normalized.counts=normExprData,
            design.matrix=designMatrix,
            gene.name="Fabp7",
            res.o=gene.map, show.plot=TRUE, plotly.flag=TRUE)

```


```{r}
source("../../R/heatmapFunctions.R")
de.genes.entr <- union(rownames(venn$int), rownames(venn$XnoY))
de.genes.entr <- union(de.genes.entr, rownames(venn$YnoX))

gene.map <- convertGenesViaMouseDb(gene.list=de.genes.entr, 
                            fromType="ENTREZID")
de.genes.symb <- attachGeneColumnToDf(as.data.frame(de.genes.entr, 
                                                    row.names=de.genes.entr), 
                                    genesMap=gene.map, 
                                    rowNamesIdentifier="ENTREZID", 
                                    mapFromIdentifier="ENTREZID",
                                    mapToIdentifier="SYMBOL")

de.genes.symb[which(is.na(de.genes.symb$gene)),]
de.genes.symb$gene[which(de.genes.symb$gene=="100039826")]  <- "Gm2444" ## not annotated in ncbi
de.genes.symb$gene[which(de.genes.symb$gene=="210541")]  <- "Entrez:210541" ## not annotated in ncbi


de.genes.counts <- normExprData[match(de.genes.symb$de.genes.entr, rownames(normExprData)),]
rownames(de.genes.counts) <- de.genes.symb$gene

de.gene.means <- computeGeneMeansOverGroups(counts=de.genes.counts, 
                            design=designMatrix, groupColumn="gcondition")

library(gplots)
color.palette = c("black", "yellow")
pal <- colorRampPalette(color.palette)(n = 1000)
# pal <- colorRampPalette(c("yellow","red"))(n = 100)
lmat = rbind(c(0,3,0), c(2,1,0), c(0,4,0))
lwid = c(1, 4, 1)
lhei = c(1, 4, 1)
row.labels.percent=0.5

filter <- apply(de.gene.means, 1, function(x) log(x[4]/x[3]) * log(x[2]/x[1]) < 0)
filter[is.na(filter)] <- FALSE
table(filter)

heatmap.2(
    as.matrix(log(de.gene.means[filter,]+1)), 
    col=pal, 
    scale="row", 
    dendrogram="row", ## manage dendrogram
    density.info="none", ## manage density in legend
    trace="none", ## manage trace along columns
    # margins=c(10, 10), ## manage margins
    Colv=NA, ## manage clustering on columns
    main=NA,
    lmat=lmat,
    lhei=lhei,
    lwid=lwid,
    #cexRow=row.labels.percent,
    colsep=c(2)
    # ,key.par = list(cex=0.5)
    # ,keysize = 0.3
    # , breaks = 5000
    # , key=FALSE
)

library(pheatmap)
filter2 <- rowMeans(de.gene.means)>0
pheatmap(log(de.gene.means[filter2,]+1), cluster_cols = FALSE, scale = "row", color = pal, border_color = NA, fontsize_row = 5)

pheatmap(log(de.gene.means[filter,]+1), cluster_cols = FALSE, scale = "row", color = pal, border_color = NA, fontsize_row = 8)

pheatmap(log(de.genes.counts[filter2,]+1), cluster_cols = FALSE, scale = "row", color = pal, border_color = NA, fontsize_row = 5, gaps_col = c(5, 10, 15))

pheatmap(log(de.genes.counts[filter,]+1), cluster_cols = FALSE, scale = "row", color = pal, border_color = NA, fontsize_row = 8, gaps_col = c(5, 10, 15))

heatmap(as.matrix(de.gene.means), Colv = NA, Rowv = NA, scale="row")


heatmap.2(
    as.matrix(log(de.genes.counts[filter,]+1)),
    col=pal, 
    scale="row", 
    dendrogram="row", ## manage dendrogram
    density.info="none", ## manage density in legend
    trace="none", ## manage trace along columns
    # margins=c(10, 10), ## manage margins
    Colv=NA, ## manage clustering on columns
    main="title",
    lmat=lmat,
    lhei=lhei,
    lwid=lwid,
    cexRow=row.labels.percent,
    colsep=c(5, 10, 15)
    # ,key.par = list(cex=0.5)
    # ,keysize = 0.3
    # , breaks = 5000
    # , key=FALSE
)

```

```{r}
# which(kowthc5 
# kowtsd5
```




